core/src/main/scala/chisel3/internal/MonoConnect.scala

```diff
@@ -372,26 +372,45 @@ private[chisel3] object MonoConnect {
   }
 
   // This function (finally) issues the connection operation
-  private def issueConnect(sink: Element, source: Element)(implicit sourceInfo: SourceInfo): Unit = {
-    // If the source is a DontCare, generate a DefInvalid for the sink,
-    //  otherwise, issue a Connect.
-    source.topBinding match {
-      case b: DontCareBinding => pushCommand(DefInvalid(sourceInfo, sink.lref))
-      case _ => pushCommand(Connect(sourceInfo, sink.lref, source.ref))
+  private def issueConnect(sink: BaseType, source: BaseType)(implicit sourceInfo: SourceInfo): Unit = {
+    (sink, source) match {
+      case (_: Element, _: Element) => {
+        // If the source is a DontCare, generate a DefInvalid for the sink,
+        //  otherwise, issue a Connect.
+        source.topBinding match {
+          case b: DontCareBinding => pushCommand(DefInvalid(sourceInfo, sink.lref))
+          case _ => pushCommand(Connect(sourceInfo, sink.lref, source.ref))
+        }
+      }
+      case (_, _) => {
+        // Internal error: issueConnect should not have been called with the given source and sink.
+        throwException(
+          s"Internal Error! Cannot issue connect for Sink (${sink.cloneType.toString}) and Source (${source.cloneType.toString})."
+        )
+      }
     }
   }
 
   // This function checks if element-level connection operation allowed.
   // Then it either issues it or throws the appropriate exception.
   def elemConnect(
     implicit sourceInfo: SourceInfo,
-    _sink:               Element,
-    _source:             Element,
+    _sink:               BaseType,
+    _source:             BaseType,
     context_mod:         RawModule
   ): Unit = {
     import BindingDirection.{Input, Internal, Output} // Using extensively so import these
-    val sink = reify(_sink)
-    val source = reify(_source)
+
+    // Reify sink and source if they're views.
+    // Views only work with Data, so for other BaseTypes, there is nothing to reify.
+    val sink = _sink match {
+      case (d: Element) => reify(d)
+      case _ => _sink
+    }
+    val source = _source match {
+      case (d: Element) => reify(d)
+      case _ => _source
+    }
     // If source has no location, assume in context module
```

Comment:
I get the reuse you're going for here, but the the holes this blows in the type system just make me think we're missing something. Changing the type to `BaseType` means one could now call this with two `Bundles` but that obviously won't do the right thing and provides fairly gnarly potential for a future developer to try to call this function with the wrong type and have potentially bad things happen.

My general view of this function is that it isn't very good and should be rewritten anyway and I think it's better to just write a new one for `Property`. That being said, we don't need to fix everything to make progress so my opinion is that if you're going to reuse this function for Property, at least make it impossible for Chisel developers to call it incorrectly by doing this following:

Please make this a package private object that is outside of `MonoConnect`. Make the main implementation of the function (what I'm commenting on here) a private internal implementation function in the new object. Then, expose one public API that takes `Element` and just calls the private implementation, `MonoConnect` will call this version. Then when you add `Property`, you can make another public version that calls the private implementation. This way, a Chisel developer won't see a function called `elemConnect` that takes a `BaseType` and try to pass a `Bundle` to it.

Reply:
Got it, that makes sense to me.

Reply:
Should this private internal implementation still call issueConnect? In the other thread I was thinking the checking could be separated from the issuing of connects.

Reply:
I think `issueConnect ` is only ever called by `elemConnect` so you can just include it as a private method in the new `private[chisel3] package elemConnect`

Reply:
Yeah I can do that. I guess I was thinking about the other thread about the changes to issueConnect. I guess I'm imagining there could be a way to re-use this for Properties without mucking with issueConnect this way. Something like this:

```
object MonoConnect {
  def elemConnect(lhs: Element, rhs: Element) = {
    checkConnects(lhs, rhs)
    issueConnect(lhs, rhs)
  }
}

object checkConnects {
  def apply(lhs: Element, rhs: Element) = {
    // call private method for checks
  }

  def apply(lhs: Property[_], rhs: Property[_]) = {
    // call private method for checks
  }
}
```

This way, the `:=` for Property can call `checkConnects`, and then just issue its own property assignment command. It gets to re-use the checks, but doesn't have to muck with issueConnect.

Reply:
The most recent push shows what I mean, curious what you think.